_posts/2020-08-25-googlectf-2020-root-power.markdown

```diff
@@ -0,0 +1,175 @@
+
+We were provided with a virtual machine disk image and had to recover the ___root password___.
+
+## Outline
+
+1. First look at disk image.
+2. Access the filesystem.
+3. Discover the authentication mechanism.
+4. Reverse engineering a kernel module.
+5. Discovering what `initramfs` contains and does.
+6. Reverse engineering an AML file.
+7. Win.
+
+## Tl;dr
+
+The given image has a pluggable authentication module that checks if a char device `/dev/chck` reads __1__ when __root__ login is attempted.
+The kernel module responsible for said device reads from ACPI, logic of which is within a ACPI Machine Language (AML) file.
+Decompiling it shows that it is a PS/2 keyboard device, and the make/break codes form the flag.
+
+## First look
+
+The archive contains a disk image and a script to run it in QEMU.
+GRUB menu reveals that it is Arch Linux, and you are greeted with a login prompt.
+
+
+![Grub Menu](/images/google20/root-power/qemu-1.png "Grub Menu")
+![Login Prompt](/images/google20/root-power/qemu-2.png "Login Prompt")
+
+Binwalk suggests there's a EXT filesystem at offset `0x100000`.
+
+![Binwalk](/images/google20/root-power/binwalk.png "Binwalk")
+
+## The filesystem
+
+The filesystem can be mounted with:
+
+```bash
+mkdir -p mnt && sudo mount -o loop,offset=1048576 disk.img mnt
+```
+
+My first instinct was to check `/etc/shadow` and the boot files, the former was a rabbit hole.
+I also checked `initramfs` to see if there are any files I might've missed, it contains `ssdt.aml`, we'll come back to this later.
+
+## Authentication mechanism
+A feature that authenticated users in Linux was PAM.
+So I checked out the modules in `/etc/pam.d/system-auth`, and `pam_chck.so` was mentioned.
+
+`pam_chck.so` looks for the __root__ user, then returns 0 (success) if `check_device() == 1`.
+
+```c
+ulong pam_sm_authenticate(undefined8 param_1)
+{
+  int r;
+  ulong ret;
+  char *user;
+  uint _r;
+  
+  _r = pam_get_user(param_1,&user,"Username: ",&user);
+  if (_r == 0) {
+    r = strcmp(user,"root");
+    if (r == 0) {
+      fwrite("Password: ",1,10,stderr);
+      r = check_device();
+      if (r == 1) {
+        fprintf(stderr,"\n\nWelcome %s\n",user);
+        ret = 0;
+      }
+      else {
+        fwrite("Wrong username or password",1,0x1a,stderr);
+        ret = 6;
+      }
+    }
+    else {
+      ret = 6;
+    }
+  }
+  else {
+    ret = (ulong)_r;
+  }
+  return ret;
+}
+```
+
+`check_device()` is a simple function that reads 2 bytes from `/dev/chck` and returns the result as an integer.
+
+```c
+int check_device(void)
+{
+  int ret;
+  char buf [2];
+  FILE *fd;
+  
+  fd = fopen("/dev/chck","r");
+  fgets(buf,2,fd);
+  fclose(fd);
+  ret = atoi(buf);
+  return ret;
+}
+```
+
+What exactly provides this device?
+
+
+## Chck kernel module
+
+Devices in `/dev/` are usually added through kernel modules, so I looked for `.ko` files in the image.
```

Comment:
I'm not sure "_usually_" is true, but I'm being nitpicky

Reply:
Ah yeah you're right, I think _"__Additional__ devices in `/dev/` are usually added through..."_ was what I meant.